zkevm-circuits/src/evm_circuit/execution/storage/sload.rs

```diff
@@ -0,0 +1,409 @@
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        self.tx_id
+            .assign(region, offset, Some(F::from(tx.id as u64)))?;
+        self.rw_counter_end_of_reversion.assign(
+            region,
+            offset,
+            Some(F::from(call.rw_counter_end_of_reversion as u64)),
+        )?;
+        self.is_persistent
+            .assign(region, offset, Some(F::from(call.is_persistent as u64)))?;
+        self.callee_address
+            .assign(region, offset, call.callee_address.to_scalar())?;
+
+        let [key, value] =
+            [step.rw_indices[4], step.rw_indices[6]].map(|idx| block.rws[idx].stack_value());
+        self.key.assign(region, offset, Some(key.to_le_bytes()))?;
+        self.value
+            .assign(region, offset, Some(value.to_le_bytes()))?;
+
+        let (_, committed_value) = block.rws[step.rw_indices[5]].aux_pair();
+        self.committed_value
+            .assign(region, offset, Some(committed_value.to_le_bytes()))?;
+
+        let (_, is_warm) = block.rws[step.rw_indices[7]].tx_access_list_value_pair();
+        self.is_warm
+            .assign(region, offset, Some(F::from(is_warm as u64)))?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct SloadGasGadget<F> {
+    is_warm: Expression<F>,
+    gas_cost: Expression<F>,
+}
+
+impl<F: Field> SloadGasGadget<F> {
+    pub(crate) fn construct(_cb: &mut ConstraintBuilder<F>, is_warm: Expression<F>) -> Self {
+        let gas_cost = select::expr(
+            is_warm.expr(),
+            GasCost::WARM_STORAGE_READ_COST.expr(),
+            GasCost::COLD_SLOAD_COST.expr(),
+        );
+
+        Self { is_warm, gas_cost }
+    }
+
+    pub(crate) fn expr(&self) -> Expression<F> {
+        // Return the gas cost
+        self.gas_cost.clone()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{
+        param::STACK_CAPACITY,
+        step::ExecutionState,
+        table::{CallContextFieldTag, RwTableTag},
+        test::{rand_fp, rand_word, run_test_circuit_incomplete_fixed_table},
+        witness::{Block, Bytecode, Call, CodeSource, ExecStep, Rw, RwMap, Transaction},
+    };
+
+    use bus_mapping::evm::OpcodeId;
+    use eth_types::{address, bytecode, evm_types::GasCost, ToWord, Word};
+    use std::convert::TryInto;
+
+    fn test_ok(
+        tx: eth_types::Transaction,
+        key: Word,
+        value: Word,
+        is_warm: bool,
+        is_persistent: bool,
+    ) {
+        let rw_counter_end_of_reversion = if is_persistent { 0 } else { 19 };
+
+        let call_data_gas_cost = tx
+            .input
+            .0
+            .iter()
+            .fold(0, |acc, byte| acc + if *byte == 0 { 4 } else { 16 });
```

Comment:
We might want to add Byte storage costs or something similar in like an `enum` form. So that these values are contained in a `const` or somewhere similar and if they change or whatever, we just need to change the constant to update them all.